### PR TITLE
Add Google benchmarks [google-benchmark]

### PR DESCRIPTION
**Pull request by camierjs**
Adds `MFEM_USE_BENCHMARK` to use [Google benchmark](https://github.com/google/benchmark) for specific CEED Bake-Off Problems and Kernels & TMOP tests.

### Installation
```
     # Check out the library.
     $ git clone https://github.com/google/benchmark.git
     # Benchmark requires Google Test as a dependency. Add the source tree as a subdirectory.
     $ git clone https://github.com/google/googletest.git benchmark/googletest
     # Go to the library root directory
     $ cd benchmark
     # Make a build directory to place the build output.
     $ cmake -E make_directory "build"
     # Generate build system files with cmake.
     $ cmake -E chdir "build" cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX="~/usr/local/benchmark" -DCMAKE_CXX_COMPILER=g++ ..
     # or, starting with CMake 3.13, use a simpler form:
     # cmake -DCMAKE_BUILD_TYPE=Release -S . -B "build"
     # Build the library.
     $ cmake --build "build" --config Release -j 12
     # Install the library
     $ cmake --install "build" --config Release --prefix ~/usr/local/benchmark
```

### CEED Bake-Off Problems and Kernels
`tests/benchmarks> ./bench_ceed --benchmark_filter=BP --benchmark_context=device=cuda`
```
---------------------------------------------------------------------------
Benchmark           Time             CPU   Iterations       MDof     MDof/s
---------------------------------------------------------------------------
BP1/1            4.21 ms         4.20 ms          162  273.659/s
BP1/2            8.68 ms         8.66 ms           81  1015.22/s
BP1/3            20.3 ms         20.3 ms           35 1.43905k/s
BP1/4            42.0 ms         41.9 ms           17 1.63843k/s
BP1/5            75.3 ms         75.2 ms            9 1.77652k/s
BP1/6             115 ms          115 ms            6 2.00573k/s
BP2/1            22.9 ms         22.8 ms           31  151.275/s
BP2/2             109 ms          108 ms            6  243.395/s
BP2/3             277 ms          277 ms            3  316.606/s
BP2/4             706 ms          704 ms            1  292.576/s
BP2/5            1277 ms         1270 ms            1  315.562/s
BP2/6            2340 ms         2328 ms            1  296.519/s
BP3/1            7.36 ms         7.35 ms           96   156.52/s
BP3/2            14.5 ms         14.5 ms           48  605.962/s
BP3/3            30.5 ms         30.5 ms           23  959.088/s
BP3/4            58.8 ms         58.6 ms           12 1.17167k/s
BP3/5             101 ms          100 ms            7 1.33126k/s
BP3/6             162 ms          162 ms            4 1.41954k/s
BP4/1            99.8 ms         99.6 ms            7  34.6469/s
BP4/2             372 ms          370 ms            2  71.2368/s
BP4/3             952 ms          949 ms            1  92.2878/s
BP4/4            2050 ms         2045 ms            1  100.763/s
BP4/5            3626 ms         3616 ms            1  110.801/s
BP4/6            6317 ms         6302 ms            1  109.515/s
BP5/1            9.89 ms         9.87 ms           67  116.525/s
BP5/2            59.6 ms         59.5 ms           12  147.759/s
BP5/3             175 ms          175 ms            4  167.361/s
BP5/4             448 ms          447 ms            2  153.673/s
BP5/5             870 ms          868 ms            1  153.834/s
BP5/6            1563 ms         1555 ms            1  147.949/s
BP6/1            26.7 ms         26.7 ms           26  129.411/s
BP6/2             186 ms          185 ms            4  142.369/s
BP6/3             548 ms          547 ms            1  160.148/s
BP6/4            1358 ms         1355 ms            1  152.094/s
BP6/5            2620 ms         2612 ms            1  153.363/s
BP6/6            4642 ms         4626 ms            1  149.198/s
```

`tests/benchmarks> ./bench_ceed --benchmark_filter=BK --benchmark_context=device=cuda`
```
---------------------------------------------------------------------------
Benchmark           Time             CPU   Iterations       MDof     MDof/s
---------------------------------------------------------------------------
BK1/1           0.087 ms        0.086 ms         8100  415.658/s
BK1/2           0.193 ms        0.193 ms         3647 1.42322k/s
BK1/3           0.399 ms        0.398 ms         1754 2.29508k/s
BK1/4           0.822 ms        0.820 ms          859 2.61658k/s
BK1/5            1.42 ms         1.42 ms          490 2.93566k/s
BK1/6            2.04 ms         2.04 ms          344 3.53261k/s
BK2/1           0.669 ms        0.667 ms         1057   161.53/s
BK2/2            3.19 ms         3.19 ms          220  258.505/s
BK2/3            7.96 ms         7.94 ms           88  344.733/s
BK2/4            20.6 ms         20.6 ms           34  312.663/s
BK2/5            37.4 ms         37.4 ms           19  335.114/s
BK2/6            68.9 ms         68.8 ms           10  313.674/s
BK3/1           0.189 ms        0.189 ms         3752  190.159/s
BK3/2           0.376 ms        0.376 ms         1856   731.29/s
BK3/3           0.715 ms        0.713 ms          987 1.28003k/s
BK3/4            1.35 ms         1.34 ms          521 1.59833k/s
BK3/5            2.23 ms         2.22 ms          311 1.87929k/s
BK3/6            3.53 ms         3.53 ms          199 2.03905k/s
BK4/1            3.07 ms         3.06 ms          229  35.1878/s
BK4/2            11.4 ms         11.3 ms           62  72.7482/s
BK4/3            29.1 ms         29.1 ms           24  94.1704/s
BK4/4            62.7 ms         62.5 ms           11  103.005/s
BK4/5             111 ms          110 ms            6  113.486/s
BK4/6             193 ms          192 ms            4  112.097/s
BK5/1           0.266 ms        0.265 ms         2636  135.519/s
BK5/2            1.78 ms         1.78 ms          392  154.623/s
BK5/3            5.23 ms         5.22 ms          134  174.937/s
BK5/4            13.5 ms         13.5 ms           52  159.592/s
BK5/5            26.3 ms         26.2 ms           27  159.249/s
BK5/6            47.2 ms         47.1 ms           15  152.634/s
BK6/1           0.787 ms        0.785 ms          895  137.351/s
BK6/2            5.58 ms         5.57 ms          126  147.933/s
BK6/3            16.5 ms         16.5 ms           43  166.348/s
BK6/4            41.0 ms         40.9 ms           17  157.367/s
BK6/5            79.1 ms         78.9 ms            9   158.67/s
BK6/6             140 ms          140 ms            5  153.878/s
```

### TMOP kernels (Kernel/D1D==Q1D/Order)
```
-------------------------------------------------------------------------------------------
Benchmark                           Time             CPU   Iterations       MDof     MDof/s
-------------------------------------------------------------------------------------------
AddMultPA/0/1                    50.5 us         50.4 us        12179  292.359/s    298.498
AddMultPA/1/1                    36.3 us         36.3 us        19488  406.435/s    418.657
AddMultPA/0/2                     100 us          100 us         6883  1076.37/s    1086.21
AddMultPA/1/2                    55.0 us         54.9 us        12475 1.96519k/s   2.00242k
AddMultPA/0/3                     193 us          192 us         3549 1.83362k/s   1.84027k
AddMultPA/1/3                     124 us          124 us         5545 2.84523k/s   2.86512k
AddMultPA/0/4                     581 us          579 us         1050 1.42261k/s   1.42209k
AddMultPA/1/4                     217 us          216 us         3162 3.80889k/s   3.82081k
AddMultGradPA/0/1                 330 us          329 us         2115  44.8042/s    44.8538
AddMultGradPA/1/1                75.0 us         74.8 us         9291  197.048/s    199.604
AddMultGradPA/0/2                 686 us          684 us          995  157.566/s     157.46
AddMultGradPA/1/2                 290 us          289 us         2421   372.96/s    373.561
AddMultGradPA/0/3                1275 us         1272 us          531   277.47/s    277.069
AddMultGradPA/1/3                 587 us          586 us         1124  602.295/s    602.075
AddMultGradPA/0/4                1995 us         1990 us          341  414.039/s    413.307
AddMultGradPA/1/4                1282 us         1279 us          526  644.181/s    643.241
GetLocalStateEnergyPA/0/1        45.8 us         45.7 us        15161  322.555/s    330.532
GetLocalStateEnergyPA/1/1        31.5 us         31.4 us        22303  468.749/s      486.1
GetLocalStateEnergyPA/0/2        78.3 us         78.1 us         8722  1.3803k/s   1.39899k
GetLocalStateEnergyPA/1/2        46.9 us         46.7 us        15031 2.30613k/s   2.36194k
GetLocalStateEnergyPA/0/3         140 us          139 us         4873 2.53162k/s   2.54787k
GetLocalStateEnergyPA/1/3        85.0 us         84.8 us         7944 4.16312k/s   4.21425k
GetLocalStateEnergyPA/0/4         259 us          258 us         2646 3.19178k/s   3.19935k
GetLocalStateEnergyPA/1/4         152 us          152 us         4490  5.4224k/s   5.45375k
AssembleGradDiagonalPA/0/1        805 us          803 us          868  18.3456/s    18.3286
AssembleGradDiagonalPA/1/1        181 us          181 us         3851  81.4798/s    81.8001
AssembleGradDiagonalPA/0/2       2108 us         2103 us          331  51.2566/s    51.1641
AssembleGradDiagonalPA/1/2        899 us          897 us          778  120.173/s    120.044
AssembleGradDiagonalPA/0/3       3916 us         3907 us          179  90.3312/s    90.1477
AssembleGradDiagonalPA/1/3       2212 us         2207 us          315  159.949/s    159.654
AssembleGradDiagonalPA/0/4       6245 us         6230 us          100  132.245/s    131.953
AssembleGradDiagonalPA/1/4       4381 us         4371 us          158  188.505/s    188.112
```

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2279](https://github.com/mfem/mfem/pull/2279) | @camierjs | @tzanio | @adrienbernede + @acfisher + @artv3 + @YohannDudouit | 08/18/21 | 09/22/21 | 09/27/21 | |

### CODE REVIEW

**Comment by camierjs** [comment]
Here are the results of the unit benchmark that mimics the `mfem::Vector` with and without `virtuals`.
```
Run on (12 X 2600 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x6)
  L1 Instruction 32 KiB (x6)
  L2 Unified 256 KiB (x6)
  L3 Unified 9216 KiB (x1)
Load Average: 1.64, 2.02, 1.98
Vector_EQ_MFEM/1                                 5.70 ns
Vector_EQ_MFEM/4                                 7.02 ns
Vector_EQ_MFEM/16                                8.27 ns
Vector_EQ_MFEM/64                                12.15 ns
Vector_EQ_MFEM/256                               36.50 ns
Vector_EQ_MFEM/1024                              134.26 ns
Vector_EQ/1                                      2.98 ns
Vector_EQ/4                                      4.11 ns
Vector_EQ/16                                     6.69 ns
Vector_EQ/64                                     6.47 ns
Vector_EQ/256                                    18.42 ns
Vector_EQ/1024                                   70.83 ns
Vector_EQ_Virtuals/1                             5.45 ns
Vector_EQ_Virtuals/4                             6.46 ns
Vector_EQ_Virtuals/16                            6.59 ns
Vector_EQ_Virtuals/64                            11.08 ns
Vector_EQ_Virtuals/256                           23.89 ns
Vector_EQ_Virtuals/1024                          72.27 ns
Vector_EQ_Virtuals_Inlines/1                     5.43 ns
Vector_EQ_Virtuals_Inlines/4                     6.40 ns
Vector_EQ_Virtuals_Inlines/16                    6.97 ns
Vector_EQ_Virtuals_Inlines/64                    12.28 ns
Vector_EQ_Virtuals_Inlines/256                   24.63 ns
Vector_EQ_Virtuals_Inlines/1024                  74.27 ns
Vector_PE_MFEM/1                                 10.11 ns
Vector_PE_MFEM/4                                 11.65 ns
Vector_PE_MFEM/16                                12.71 ns
Vector_PE_MFEM/64                                21.87 ns
Vector_PE_MFEM/256                               57.71 ns
Vector_PE_MFEM/1024                              211.50 ns
Vector_PE_MFEM_64/1                              70.27 ns
Vector_PE_MFEM_64/4                              70.26 ns
Vector_PE_MFEM_64/16                             71.54 ns
Vector_PE_MFEM_64/64                             78.11 ns
Vector_PE_MFEM_64/256                            114.23 ns
Vector_PE_MFEM_64/1024                           273.27 ns
Vector_PE/1                                      5.16 ns
Vector_PE/4                                      6.47 ns
Vector_PE/16                                     7.83 ns
Vector_PE/64                                     15.29 ns
Vector_PE/256                                    57.39 ns
Vector_PE/1024                                   202.19 ns
Vector_PE_Virtuals/1                             8.60 ns
Vector_PE_Virtuals/4                             9.58 ns
Vector_PE_Virtuals/16                            11.05 ns
Vector_PE_Virtuals/64                            18.69 ns
Vector_PE_Virtuals/256                           53.74 ns
Vector_PE_Virtuals/1024                          204.07 ns
Vector_Virtuals_Inlines_PE/1                     8.25 ns
Vector_Virtuals_Inlines_PE/4                     9.63 ns
Vector_Virtuals_Inlines_PE/16                    10.92 ns
Vector_Virtuals_Inlines_PE/64                    18.47 ns
Vector_Virtuals_Inlines_PE/256                   53.77 ns
Vector_Virtuals_Inlines_PE/1024                  201.43 ns
Base/1                                           0.25 ns
Base/4                                           0.25 ns
Base/16                                          0.25 ns
Base/64                                          0.25 ns
Base/256                                         0.25 ns
Base/1024                                        0.25 ns
Base_Virtuals_inlined/1                          0.25 ns
Base_Virtuals_inlined/4                          0.25 ns
Base_Virtuals_inlined/16                         0.25 ns
Base_Virtuals_inlined/64                         0.25 ns
Base_Virtuals_inlined/256                        0.25 ns
Base_Virtuals_inlined/1024                       0.25 ns
Base_Virtuals_Derived_not_inlined/1              1.24 ns
Base_Virtuals_Derived_not_inlined/4              1.26 ns
Base_Virtuals_Derived_not_inlined/16             1.24 ns
Base_Virtuals_Derived_not_inlined/64             1.24 ns
Base_Virtuals_Derived_not_inlined/256            1.24 ns
Base_Virtuals_Derived_not_inlined/1024           1.25 ns

**Comment by tzanio** [comment]
Thanks for running these @camierjs!

Can you explain the different labels? For example what is the difference between

```
Vector_EQ_MFEM/4                                 7.02 ns
Vector_EQ/4                                      4.11 ns
Vector_EQ_Virtuals/4                             6.46 ns
Vector_EQ_Virtuals_Inlines/4                     6.40 ns
```

(I assume all of these are for `operator=` with vectors of size 4)

**Comment by tzanio** [comment]
Also: nicely done with the support for Google benchmark 👍

**Comment by camierjs** [comment]
`EQ` is used for the `Operator=` and `PE` for `Operator+=`.
```
Vector_EQ_MFEM/4              benchmark using the real `mfem::Vector` class, which now uses `virtuals`
Vector_EQ/4                   minimal/trimmed `Vector` class without any `virtuals`
Vector_EQ_Virtuals/4          minimal `Vector` class with the two `virtuals` for the `=` and `+=` operators
Vector_EQ_Virtuals_Inlines/4  minimal `Vector` class with the two `virtuals` and `inline`d functions
```
`/n` is for the size, yes.

**Comment by camierjs** [comment]
There seems to be no hit in performance for the `Base/n` vs. `Base_Virtuals_inlined/n`, compared to when we need to use the `vtable`.

However, it's not the case for the `Vector` classes.

**Comment by camierjs** [comment]
LGTM

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: config/defaults.cmake [hunk 1]
```diff
@@ -74,6 +75,7 @@ set(MFEM_MPI_NP 4 CACHE STRING "Number of processes used for MPI tests")
 option(MFEM_ENABLE_TESTING "Enable the ctest framework for testing" ON)
 option(MFEM_ENABLE_EXAMPLES "Build all of the examples" OFF)
 option(MFEM_ENABLE_MINIAPPS "Build all of the miniapps" OFF)
+option(MFEM_ENABLE_BENCHMARKS "Build all of the benchmarks" OFF)
```
Comment (artv3) [inline config/defaults.cmake:78]: To be more clear, what about "MFEM_ENABLE_GOOGLE_BENCHMARKS"

Comment (camierjs) [reply config/defaults.cmake]: `done`

**Comment by camierjs** [comment]
LGTM

**Comment by camierjs** [comment]
@artv3 , @adrienbernede & @acfisher, this should be ready for review: I've included the installation procedure and added all the CEED benchmarks (BP & BK).

File: tests/benchmarks/bench_ceed.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      cg.Mult(B,X);
+      MFEM_DEVICE_SYNC;
+      mdofs += MDofs() * cg.GetNumIterations();
+   }
+};
+
+/// Bake-off Problems (BPs)
+#define BakeOff_Problem(i,Kernel,VDIM,p_eq_q)\
+static void BP##i(bm::State &state){\
+   Problem<Kernel##Integrator,VDIM,p_eq_q> ker(state.range(0));\
```
Comment (adrienbernede) [inline tests/benchmarks/bench_ceed.cpp:116]: Where does `##Integrator` comes from?

File: fem/bilininteg.hpp [hunk 3]
```diff
@@ -1982,24 +1982,32 @@ class DiffusionIntegrator: public BilinearFormIntegrator
 
 public:
    /// Construct a diffusion integrator with coefficient Q = 1
-   DiffusionIntegrator()
-      : Q(NULL), VQ(NULL), MQ(NULL), SMQ(NULL), maps(NULL), geom(NULL) { }
+   DiffusionIntegrator(const IntegrationRule *ir = nullptr)
+      : BilinearFormIntegrator(ir),
+        Q(NULL), VQ(NULL), MQ(NULL), SMQ(NULL), maps(NULL), geom(NULL) { }
```
Comment (adrienbernede) [inline fem/bilininteg.hpp:1987]: I struggle to see how are those changes related to the addition of benchmark?

File: INSTALL [hunk 4]
```diff
@@ -516,6 +520,9 @@ MFEM_USE_CALIPER = YES/NO
    profiling at runtime with Caliper's configuration API. Alternatively, one
    can configure Caliper through environment variables or config files.
 
+MFEM_USE_BENCHMARK = YES/NO
+   Enables the google-benchmarks in the tests/benchmarks directory.
+
```
[the remark this answers is not in the document]
Comment (camierjs) [reply INSTALL]: `done`

File: tests/benchmarks/bench_ceed.cpp (hunk 2 above)
Comment (camierjs) [reply tests/benchmarks/bench_ceed.cpp:116]: It's a [concatenation](https://gcc.gnu.org/onlinedocs/cpp/Concatenation.html) with the `Kernel` name.

Comment (adrienbernede) [reply tests/benchmarks/bench_ceed.cpp:116]: I see, so "Integrator" here wont be substituted and ## is just a separator to have "Kernel" substituted and be able to get something like "DiffusionIntegrator" in the end, right?

**Comment by adrienbernede** [comment]
LGTM

**Review by adrienbernede — APPROVED** [review]

**Review by YohannDudouit — APPROVED** [review]
I love it! Thanks @camierjs for doing this, this is very useful! We should soon be able to do performance tracking with this!

I tried it with libCEED, it also works perfectly:
```
[dudouit1@lassen522:benchmarks]$ ./bench_ceed --benchmark_filter=BP --benchmark_context=device=ceed-cuda
device : ceed-cuda
Device configuration: ceed-cuda,cuda,cpu
libCEED backend: /gpu/cuda/gen
Memory configuration: host-std,cuda
2021-09-21T14:32:14-07:00
Running ./bench_ceed
Run on (176 X 3450 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x44)
  L1 Instruction 32 KiB (x44)
  L2 Unified 512 KiB (x22)
  L3 Unified 10240 KiB (x22)
Load Average: 1.31, 11.46, 14.69
device: ceed-cuda
----------------------------------------------------------------
Benchmark           Time             CPU   Iterations     MDof/s
----------------------------------------------------------------
BP1/1            3.65 ms         3.65 ms          190  315.221/s
BP1/2            5.29 ms         5.29 ms          126 1.66039k/s
BP1/3            11.4 ms         11.4 ms           61 2.55268k/s
BP1/4            21.6 ms         21.6 ms           32 3.18714k/s
BP1/5            37.9 ms         37.9 ms           18 3.52154k/s
BP1/6            61.6 ms         61.6 ms           11 3.73346k/s
BP2/1            4.71 ms         4.71 ms          153  732.041/s
BP2/2            11.0 ms         11.0 ms           63 2.39608k/s
BP2/3            27.8 ms         27.8 ms           25 3.15347k/s
BP2/4            69.5 ms         69.5 ms           10 2.96642k/s
BP2/5             211 ms          211 ms            3 1.89776k/s
BP2/6             309 ms          309 ms            2 2.23619k/s
BP3/1            5.14 ms         5.14 ms          134  223.699/s
BP3/2            8.42 ms         8.42 ms           82  1043.77/s
BP3/3            18.5 ms         18.5 ms           38 1.57993k/s
BP3/4            34.5 ms         34.5 ms           20 1.99273k/s
BP3/5            58.5 ms         58.5 ms           11 2.28271k/s
BP3/6            92.4 ms         92.4 ms            7 2.48959k/s
BP4/1            7.60 ms         7.60 ms           91  453.765/s
BP4/2            17.0 ms         17.0 ms           41 1.55525k/s
BP4/3            52.1 ms         52.1 ms           12 1.68142k/s
BP4/4             115 ms          115 ms            6 1.78486k/s
BP4/5             366 ms          366 ms            2  1095.92/s
BP4/6             478 ms          477 ms            2  1.4455k/s
BP5/1            3.80 ms         3.80 ms          184  302.345/s
BP5/2            6.60 ms         6.60 ms          100 1.33156k/s
BP5/3            13.7 ms         13.7 ms           51 2.13678k/s
BP5/4            27.5 ms         27.5 ms           25 2.49823k/s
BP5/5            49.7 ms         49.7 ms           13 2.68792k/s
BP5/6            82.0 ms         82.0 ms            8 2.80418k/s
BP6/1            5.00 ms         5.00 ms          134  689.567/s
BP6/2            14.2 ms         14.2 ms           49 1.86215k/s
BP6/3            29.9 ms         29.9 ms           23 2.93458k/s
BP6/4            78.0 ms         78.0 ms            9 2.64243k/s
BP6/5             190 ms          190 ms            4 2.10732k/s
BP6/6             383 ms          383 ms            2 1.80214k/s
```

Interestingly, we can see that using libCEED results in substantial speedups for all BPs on Lassen:
```
BP1 ~2x
BP2 ~8-10x
BP3 ~1.5-2x
BP4 ~15-20x
BP5 ~15-20x
BP6 ~15-20x
```

**Review by artv3 — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
@camierjs can you help with merging `master` and resolving the conflicts?

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Review by tzanio — APPROVED** [review]
This is so cool, thanks for adding it @camierjs!!

👍 🚀 🎖️

**Pull request merged**